Paywall Block: Appearance related to Newsletter Visibility

lezama opened the pull request:
Partially #32124 

Make the Paywall Block change in the editor depending on Newsletter Visibility

## Proposed changes:

### Disabled
<img width="887" alt="Screenshot 2023-08-01 at 15 46 48" src="https://github.com/Automattic/jetpack/assets/104869/4d7ca0de-d159-4ae0-bf5b-4cbfd99d0057">

### Anyone Subscribed
<img width="887" alt="Screenshot 2023-08-01 at 15 46 38" src="https://github.com/Automattic/jetpack/assets/104869/be6db5d9-39a4-4e0f-b51a-854ad378e098">

### Pair Subscribers
<img width="889" alt="Screenshot 2023-08-01 at 15 46 27" src="https://github.com/Automattic/jetpack/assets/104869/edd1a0d7-346f-46b2-9643-f85f5b66c1b6">

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Add a Paywall block.
* Check the appearance changes whenever you change the Newsletter Visibility settings.
* Make sure paywall is still working.

lezama committed:
[not verified] dynamic copy

lezama committed:
[not verified] changelog

crisbusquets commented:
Hi @lezama !

### Disabled
`Change visibility to enable paywall`

<img width="1341" alt="Screenshot 2023-08-02 at 15 27 00" src="https://github.com/Automattic/jetpack/assets/25607244/66cf165f-afe4-464d-92f2-05c03d4fc0b4">


### Anyone subscribed
`Subscriber-only content below`

<img width="1324" alt="Screenshot 2023-08-02 at 15 27 18" src="https://github.com/Automattic/jetpack/assets/25607244/52261029-5180-4cdd-8227-9d7e456e4205">


### Paid subscribers
`Paid content below this line`

<img width="1347" alt="Screenshot 2023-08-02 at 15 27 37" src="https://github.com/Automattic/jetpack/assets/25607244/025e6368-b5ed-4e66-ac4b-cb480ac96223">

[Figma file](qQxuiiWjgNV0AiYg6070eo-fi-5_3941).

simison commented on projects/plugins/jetpack/extensions/blocks/paywall/edit.js:
```diff
@@ -1,15 +1,30 @@
+	const text =
+		accessLevel === accessOptions.subscribers.key
+			? __( 'Subscribers Only', 'jetpack' )
+			: __( 'Paywall', 'jetpack' );
```
I think this would be good use of `switch` statement to switch between the 3 labels. :-)

simison commented on projects/plugins/jetpack/extensions/blocks/subscriptions/utils.js:
```diff
@@ -69,3 +71,16 @@ export default function GetAddPaidPlanButton( { context = 'other', hasNewsletter
 	);
 }
+
+export const GetAccessLevel = postType => {
```
`getAccessLevel` since not a component.

simison replied:
could be good use for a hook https://react.dev/learn/reusing-logic-with-custom-hooks

lezama replied:
> getAccessLevel since not a component.

it was like that originally but ESLint complained about using `useEntityProp` inside a non component 😅 

> could be good use for a hook https://react.dev/learn/reusing-logic-with-custom-hooks

👀

edanzer commented on projects/plugins/jetpack/extensions/blocks/paywall/edit.js:
```diff
@@ -1,15 +1,30 @@
+import { store as editorStore } from '@wordpress/editor';
+import { __ } from '@wordpress/i18n';
+import { accessOptions } from '../subscriptions/constants';
+import { GetAccessLevel } from '../subscriptions/utils';
```
Given that `accessOptions` and `getAccessLevel` are used by multiple blocks (here and subscriptions), I wonder if those should be moved to someplace more general rather than importing from one block to another? Not sure where. Maybe /project/plugins/jetpack/extensions/shared?

simison replied:
Yeah, I think we should move a bunch of shared code away from `subscriptions` to someplace other, and structure things under subscription form better, too.

We could start by having this under `shared/subscriptions` or similar?

lezama replied:
`shared/newsletters` 😄

lezama replied:
I was lazy to create a new file and moved all common methods to shared/memberships 😅

edanzer commented:
This tests as expected for me. I left one comment. I'll plan to test again once merge conflicts and other feedback above are addressed.

lezama committed:
[not verified] use switch and better copies

lezama committed:
[not verified] refactor to shared

lezama commented:
> This tests as expected for me. I left one comment. I'll plan to test again once merge conflicts and other feedback above are addressed.

@edanzer I addressed all the feedback and rebased. Since I moved a lot of code to `shared/memberships` could you take another look? 🙇

simison commented on projects/plugins/jetpack/extensions/blocks/subscriptions/panel.js:
```diff
@@ -281,18 +285,13 @@ function NewsletterPostPublishSettingsPanel( { accessLevel, isModuleActive } ) {
+
 	const { tracks } = useAnalytics();
 
 	// Set the accessLevel to "everybody" when one is not defined
```
The comment seems unnecessary since it explains logic happening inside the hook

simison commented on projects/plugins/jetpack/extensions/shared/memberships.js:
```diff
@@ -88,3 +92,40 @@ export const reloadPageWithPremiumContentQueryString = function (
 	}
 	document.location.href = newQueryString;
 };
+
```
Since the `shared/memberships.js` is used by `view.js` options, it sets some limitations as how to write JS and avoid importing dependencies.

To keep that boundary clearer, let's split the file to `shared/memberships-view.js` and `shared/memberships-edit.js`?

simison commented on projects/plugins/jetpack/extensions/shared/memberships.js:
```diff
@@ -1,5 +1,9 @@
 
+import { getJetpackData } from '@automattic/jetpack-shared-extension-utils';
+import { useEntityProp } from '@wordpress/core-data';
+import { __ } from '@wordpress/i18n';
```
I noticed a [warning](https://github.com/Automattic/jetpack/actions/runs/5742658272/job/15565234375?pr=32201#step:9:5138) that I'm not sure was there before:

```
[plugins/jetpack 3:12.045] [build-production-extensions] Block view script dependency check detected issues!

[plugins/jetpack 3:12.045] [build-production-extensions] In projects/plugins/jetpack/_inc/blocks/subscriptions/view.asset.php
[plugins/jetpack 3:12.045] [build-production-extensions] Dependency `lodash` should not be used by the subscriptions block's view.js.
[plugins/jetpack 3:12.045] [build-production-extensions] Dependency `wp-compose` should not be used by the subscriptions block's view.js.
[plugins/jetpack 3:12.045] [build-production-extensions] Dependency `wp-core-data` should not be used by the subscriptions block's view.js.
[plugins/jetpack 3:12.045] [build-production-extensions] Dependency `wp-element` should not be used by the subscriptions block's view.js.
[plugins/jetpack 3:12.045] [build-production-extensions] Dependency `wp-plugins` should not be used by the subscriptions block's view.js.

[plugins/jetpack 3:12.045] [build-production-extensions] Dependencies allowed may be adjusted by editing the arrays at the top of projects/plugins/jetpack/tools/check-block-assets.php.
```

jeherve replied:
That would definitely be a blocker, we do not want to increase the size of the frontend's bundle.

simison replied:
I would've expected build process to get rid of thoes for the `view` since any of the imported functions aren't actually used there, but it depends how the webpack is configured and if "tree shaking" applies.

simison reviewed (dismissed):
Technically works already and I'm happy with the code.

Just small change to split the shared file into two would be good before merging.

Note the failing build CI task.

lezama committed:
add shared/memberships-edit.js

lezama committed:
clean old comment

lezama committed:
fix imports

lezama commented:
@jeherve @simison moved the common methods to a memberships-edit.js file, looks like the warning went away
